### ISSUE TITLE
Single main content media item with multiple content breaks containing multiple ads in each. 

### ISSUE THREAD

**Issue by chris-sloan** [google/ExoPlayer#9048]
I'm trying to update our implementation of Exoplayer moving from manually handling VAST ads to using the IMA Extension, with MediaItems and a compiled playlist. 

I'm having trouble playing midroll and postroll ads. 
I've come across [this issue](https://github.com/google/ExoPlayer/issues/3694#issuecomment-365627568) from 2018 where it says there is little demand for handling VAST ads at specified points in playback. 

Our use case, is pre, mid and post roll ads, multiple midroll content breaks and almost definitely multiple ads per content break. Even though I know about (most of) the ads in advance, there can sometimes be 40+ and so having to parse through each to manually build a VMAP response (as per the suggested 'workaround') prior to commencing playback isn't ideal. 

I'm wondering, is this still the case? Has there been any advancement? 

Can I make some attempt to contribute to add this feature? Where should I start looking? 

Are there any other suggestions that could help me out? 

Thanks

**Comment by andrewlewis** [comment]
There hasn't been any advancement on the request to load a VAST ad tag and specify where it should play within the content (they still just play as prerolls).

VMAP seems like a perfect fit for what you are describing. If there's a reason not to use it then perhaps you could give some details here, and we can take that into account for prioritization.

**Comment by andrewlewis** [comment]
Just to add: we do now support playing playlists of pieces of content where each one has an ad tag, which didn't used to be supported. That might give another way to do a workaround for this (you could split up the content into clips and then specify the VAST ad tags with each one) but it seems hacky.

**Comment by chris-sloan** [comment]
Thanks @andrewlewis for the response. 

I looked into VMAP but our app supports currently vast 2 and vast 4.1. It seems according to the VMAP docs over on [iab.com](https://www.iab.com/guidelines/vmap/) that VMAP only supports up to vast 3. Also we don't really know in advance what version of the VAST ad we're going to get, so need to be really dyanmic about it. 

I've spent much of the day trying two things.

The first making an attempt to build a VMAP locally. This didn't really work for me. I couldn't determine why not initially and then discovered about only supporting vast 3 so gave up.

The second being to build multiple `MediaItems` with the same content url and add an `AdTagUri` to each, based on the number of contentBreaks and adverts required to play. In almost the same way as you describe but without actually splitting the content into clips.

```
val builder = MediaItem.Builder()
          .setUri(mediaUrl)
          .setDrmPlayClearContentWithoutKey(true)
          .setDrmSessionForClearPeriods(true)
          .setDrmLicenseUri(licenseKeyUrl)
          .setDrmUuid(C.WIDEVINE_UUID)

contentBreaks?.forEachIndexed { breakIndex, contentBreak ->
      contentBreak.adverts.forEachIndexed { adIndex, advert ->
        val builder = getDefaultMediaItemBuilder(uri, licenseKeyUrl)
            .setAdTagUri(advert.url)
            .setTag(advert)
            .setClipEndPositionMs(contentBreak.timecode)
        mediaItems.add(builder.build())
```

And this works for pre-rolls, they all play as expected, and the first mid roll hits and plays the first ad but then stops. This first mid roll ad doesn't show any of the overlay that the pre-rolls show, so I'm guessing this is some sort of unspecified behaviour.

I've played around with various settings around the `.setClipEndPositionMs(contentBreak.timecode)` and `.setClipStartPositionMs(contentBreak.timecode)` to essentially try to break the content up thinking that as each `MediaItem` only plays pre-rolls then it might work but so far to no avail. 

I can't really go back to the content supplier and ask them to break up all the content into chunks for me, that's just not feasible. 

My last resort (I think) is to continue to manually parse the VAST tags and get the video url's then build a multi item playlist entirely. I was hoping to move away from the manual side of things, we've got viewability and other tracking url's to hit at various points throughout each ad and my preference would be to offload that.

**Comment by andrewlewis** [comment]
> I looked into VMAP but our app supports currently vast 2 and vast 4.1. It seems according to the VMAP docs over on iab.com that VMAP only supports up to vast 3. Also we don't really know in advance what version of the VAST ad we're going to get, so need to be really dyanmic about it.

Do you have a server that is capable of producing a VMAP ad tag? If so I would try just turning that on and see what happens, despite the documented version limitation. You could also ask at https://groups.google.com/g/ima-sdk as they are likely to have more expertise there.

> And this works for pre-rolls, they all play as expected, and the first mid roll hits and plays the first ad but then stops. This first mid roll ad doesn't show any of the overlay that the pre-rolls show, so I'm guessing this is some sort of unspecified behaviour.

It could be a bug but it's hard to say. If you debug this and find a root cause for playback being stuck please let us know here! If you can prepare a minimal reproducible sample and share a link to that there's a chance we will look into it at some point but probably not for a while given that this is not likely to be a common usage pattern.

**Comment by chris-sloan** [comment]
>Do you have a server that is capable of producing a VMAP ad tag? 

No, and the content provider said they'll not support it in the future either. (I don't know why).

>You could also ask at https://groups.google.com/g/ima-sdk 

Good idea, will pursue this. 

>If you debug this and find a root cause for playback being stuck please let us know here! 

Will do. Thanks.

**Comment by andrewlewis** [comment]
Given your server can't produce VMAP this won't be much help but in case someone else finds this issue: I checked with the IMA developer relations team and they clarified that the IMA SDK VMAP handling is not limited to VAST 3 (it should also work with VAST 4). On the serving side, just as an example, it appears that Google Ad Manager supports this (search [this page](https://support.google.com/admanager/table/9749596?hl=en) for `xml_vmap1_vast4`).

**Comment by chris-sloan** [comment]
In actual fact, the one and only [open PR](https://github.com/InteractiveAdvertisingBureau/vmap/pull/1) for the VMAP spec is to add vast4 to the `templateType` which is where I thought my issue lay. Perhaps I should look more into actually compiling my own VMAP string. 

Thanks for your time @andrewlewis. Closing now.

**Comment by chris-sloan** [comment]
I have worked on a reproducible sample to highlight what I was trying to do, which can be found [here](https://github.com/chris-sloan/ExoPlayer/tree/multiMediaItemsForAds). 

> Just to add: we do now support playing playlists of pieces of content where each one has an ad tag, which didn't used to be supported. 

Am I missing something?, because although I give each item an ad tag, the ad is not always played. I've given an ad tag to each chunk of content but can only see the first ad being played. 

There are slight differences between how my own project handles this and how the demo project handles it, I'm guessing it's to do with exoplayer versions (we're running 2.13.2 currently). 

The functionality is obviously not supported in this way currently even though its possible to give each playlist item an ad tag. There are other issues too such as the overall content duration is not specified( because its separate items each with their own clip values, so to be expected).

**Comment by andrewlewis** [comment]
> Am I missing something?, because although I give each item an ad tag, the ad is not always played. I've given an ad tag to each chunk of content but can only see the first ad being played.

Hm, I think this could be due to the media IDs and ad tags matching, causing the player to keep just one ad playback state (so for the next item it looks like it has already been played). Please could you try passing a different `adsId` with each ad tag URI? See https://exoplayer.dev/ad-insertion.html for more information.

**Comment by chris-sloan** [comment]
Hmm, that doesn't seem to have had an effect. 
I've tried using both the uri string as the id and a id common for each pod of ads. 

Also I'm now not sure that it would ever work as the same main content is played over and over if there are multiple ads per break. eg after the first midroll i see 30s - 60s content twice (as I have specified 2 ads per break).

**Comment by andrewlewis** [comment]
Thanks for providing the sample code.

I tried removing the inner loop where the second ad tag for each content break is added and I saw what looks like the expected behavior (after each clip finishes an ad plays). This is working as expected because each iteration of the inner loop uses the same ads ID, so when the second item plays the ads are treated as already having played.

Currently it's not possible to have multiple ad tags associated with one media item, but another hacky workaround for this would be to use a short `SilenceMediaSource` as content for the first ad tag (only the second ad tag gets the real content). I think you'd need to switch to using the `MediaSource`-based player methods to do that or pass a custom `MediaSourceFactory` that lets you instantiate the `SilenceMediaSource`.
